libensemble/libE.py

```diff
@@ -204,31 +216,26 @@ def libE_mpi(sim_specs, gen_specs, exit_criteria,
     if libE_specs['comm'] == mpi_comm_null:
         return [], persis_info, 3  # Process not in comm
 
-    comm = libE_specs['comm'].Dup()
-    rank = comm.Get_rank()
-    is_master = (rank == 0)
     check_inputs(libE_specs, alloc_specs, sim_specs, gen_specs, exit_criteria, H0)
 
-    exctr = Executor.executor
-    if exctr is not None:
-        local_host = socket.gethostname()
-        libE_nodes = set(comm.allgather(local_host))
-        exctr.add_comm_info(libE_nodes=libE_nodes, serial_setup=is_master)
-
-    # Run manager or worker code, depending
-    if is_master:
-        H, persis_info, exit_flag = libE_mpi_manager(comm, sim_specs,
-                                                     gen_specs, exit_criteria,
-                                                     persis_info, alloc_specs,
-                                                     libE_specs, H0)
-    else:
+    with DupComm(libE_specs['comm']) as comm:
+        rank = comm.Get_rank()
+        is_master = (rank == 0)
+
+        exctr = Executor.executor
+        if exctr is not None:
+            local_host = socket.gethostname()
+            libE_nodes = set(comm.allgather(local_host))
+            exctr.add_comm_info(libE_nodes=libE_nodes, serial_setup=is_master)
+
+        # Run manager or worker code, depending
+        if is_master:
+            return libE_mpi_manager(comm, sim_specs, gen_specs, exit_criteria,
+                                    persis_info, alloc_specs, libE_specs, H0)
+
         # Worker returns a subset of MPI output
         libE_mpi_worker(comm, sim_specs, gen_specs, libE_specs)
```

Comment:
Is `worker_main` not returning anything? Can the `persis_info` in the next line be updated from `worker_main`?

Reply:
``worker_main`` in it's current form just initiates worker instances. ``persis_info`` is first communicated to a worker in a unit of work through the communicator, and back to the manager through the communicator as well, in a loop initialized within ``worker_main`` but without anything returned to it.

Reply:
Yes, the worker does not return anything. It is merely to satisfy the interface. Perhaps it should return an empty dictionary?

Reply:
Sounds good